design/secrets_provider_milestone1_design.md

```diff
@@ -0,0 +1,474 @@
+
+*Decision*: Solution #1, use **namespace** and **service account** to define the Secrets Provider host in Conjur policy.
+Recommend to the customer to create a dedicated service account for Secrets Provider.
+
+#### Customer Experience
+
+1. Configure K8s Secrets with ***conjur-map\*** metadata defined 
+
+   - *Optional:* add label to K8s Secret
+
+2. Create Service Account for Secrets Provider 
+
+3. Create Role for the Service Accounts with **get**/**update** and also **list** privileges on K8s Secrets resources
+
+4. Create RoleBinding and bind the Service Accounts to the Role from previous step
+
+5. Create/Deploy Secrets Providers as Jobs in deployment manifest(s)
+   - If label filtering is defined, add "K8S_SECRET_LABEL" in the Secrets Provider manifest(s)
+
+The Secrets Provider host will resemble the following:
+
+```yaml
+- !policy
+  id: secrets-accessors
+  body:
+  - !host
+    id: secrets-provider-1
+    annotations:
+      authn-k8s/namespace: prod-namespace
+      authn-k8s/service-account: prod-sa
+      authn-k8s/authentication-container-name: cyberark-secrets-provider
+
+  - !group 
+
+  - !permit
+    resources: !variable super_secret
+    role: !group secrets-accessors
+    privileges: [ read, execute ]
+ 
+- !grant
+  role: !group secrets-accessors
+  members:
+    - !host conjur/authn-k8s/secrets-provider-1
+```
+
+The Secrets Provider manifest will be defined as the following:
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: secrets-provider-job
+spec:
+  template:
+    spec:
+      serviceAccountName: prod-sa
+      containers:
+      - image: secrets-provider:latest
+        name: cyberark-secrets-provider-1
+
+# Rest of environment variables for Pod to run 
+```
+
+We recommend that the admin will define `Role` and not `ClusterRole` to follow the principle of least privilege on resources.
+
+#### Code changes
+
+In the current solution, there are behaviors that need to be changed:
+
+*Batch retrieval*
+
+At current, the Secrets Provider code looks at the "K8S_SECRET" environment variable in each of the pod manifests and preforms a batch retrieval against the Conjur API to get the relevant Conjur Secrets. If there is a failure in retrieving one of the secrets, the request fails and no secrets are returned to the request Pod. 
+
+This is problematic for us at this stage because now the Secrets Provider is sitting separately and will need to perform a batch retrieval of all the secrets in the namespace. In other words, if the Secrets Provider fails to fetch one Conjur secret, the whole request will fail and applications will not get their secrets. 
+
+To avoid this, we will parse over each K8s Secret and perform a batch retrieval request. That way, if there is a failure in retrieving a Conjur secret (a failure in permissioning for example), it will be contained to that secret and not impact all the secrets in the namespace.
```

Comment:
i wrote this before but the comment disappeared. I think that this change will cost us in performance. If a customer has 1000 k8s secrets then instead of sending just 1 batch retrieval request (like we do today) we will now send 1000 requests. This is a huge cost to fix the issue you are presenting here.

I would suggest to do one of the following:
1. Continue as today where the whole request will fail and the customer will need to check the logs to see what failed.
2. Run the request as today. If the response is not 200 check the response body to see which secrets failed to be retrieved (not sure if this data is in the response - need to verify). Log the failing secret names in WARN and filter them out from the original request. Now send the request again (it should pass now) and retrieve the rest of the secrets.

wdyt?

Reply:
Failing on _all_ k8s secrets because one failed due to permissioning is wonky UX because it means no application can operate. 
I don't know what the max number of params we can add to a batch retrieval call but let's say if we are able to send 1000... that is a lot of load on Conjur.

Currently, only the last secret that failed is beginning logged. So if we go with #2 we would need to enhance this experience anyways. For your second idea would a new request be sent from scratch just without the failed secret or would it work in a substring manner, just running a batch request from the point of the failure and on?
@abrahamko what do you think about the 2nd option?

Reply:
i think that a 1000 requests with 1 secret adds more load on conjur than 1 request with a 1000 secrets. That's the whole point of the batch retrieval. I am strongly against this suggestion.

what do you mean by `only the last secret that failed is beginning logged`? where is it logged? can you share the response you get on a batch retrieval request that has 4 secrets and 2 are not successful?

this is the code of `batch`:
```
  def batch
    variables = Resource.where(resource_id: variable_ids).eager(:secrets).all

    unless variable_ids.count == variables.count
      raise Exceptions::RecordNotFound,
            variable_ids.find { |r| !variables.map(&:id).include?(r) }
    end
    
    result = {}

    authorize_many variables, :execute
    
    variables.each do |variable|
      unless (secret = variable.last_secret)
        raise Exceptions::RecordNotFound, variable.resource_id
      end
      
      result[variable.resource_id] = secret.value
      audit_fetch variable
    end

    render json: result
  end
```

It seems that we return ALL the variable names that do not exist if such exist but only one variable name if it doesn't have value. this can change quite easily to return all the failing variables also in case they all exist but some have no value.

Reply:
@orenbm, thanks for getting so low-leveled on this.
You are right regarding the list of missing variables.
But there is another option to fail, if one is missing rights.
Then, it will get `Forbidden` without any more info, not matter how many failed.
And I think security-wise, this is the correct behavior, so there's nothing to fix here.
So in the `Forbidden` usecase, you simply cannot do what you suggested.

One more point to consider is scalability.
If you have 1000 secrets, you will send one GET request with **huge** URL.
Moreover, the server handles them all in one thread, while the client is idle waiting for response (which might fail for timeout, in case it takes too long to get them all).

On the other hand, splitting the requests to many small ones allows better concurrent handling, and can make use of followers' load-balancers.
Also, it allows concurrent handling in the client, so it doesn't have to wait for all responses to return in order to update the k8s secrets.

Reply:
i would like to get @oburstein-hub 's opinion on this. @jvanderhoof can be of an assistant as well.